pkg/asset/machines/aws/instance_types.go

```diff
@@ -23,7 +24,50 @@ func PreferredInstanceType(ctx context.Context, meta *awsconfig.Metadata, types
 		return types[0], err
 	}
 
-	client := ec2.New(sess, aws.NewConfig().WithRegion(meta.Region))
+	found, err := getInstanceTypeZoneInfo(ctx, sess, meta.Region, types, zones)
+	if err != nil {
+		return types[0], err
+	}
+
+	reqZones := sets.NewString(zones...)
+	for _, t := range types {
+		if reqZones.Difference(sets.NewString(found[t]...)).Len() == 0 {
+			return t, nil
+		}
+	}
+
+	return types[0], errors.New("no instance type found for the zone constraint")
+}
+
+// FilterZonesBasedOnInstanceType return a filtered list of zones where the particular instance type is available. This is mainly necessary for ARM, where the instance type m6g is not
+// available in all availability zones.
+func FilterZonesBasedOnInstanceType(ctx context.Context, meta *awsconfig.Metadata, instanceType string, zones []string) ([]string, error) {
+	sess, err := meta.Session(ctx)
+	if err != nil {
+		return zones, err
+	}
+
+	types := []string{instanceType}
+	found, err := getInstanceTypeZoneInfo(ctx, sess, meta.Region, types, zones)
+	if err != nil {
+		return zones, err
+	}
+
+	reqZones := sets.NewString(zones...)
+	if reqZones.Difference(sets.NewString(found[instanceType]...)).Len() == 0 {
+		return zones, nil
+	}
+
+	for k := range reqZones.Difference(sets.NewString(found[types[0]]...)) {
+		reqZones.Delete(k)
+	}
+	return reqZones.List(), nil
+}
+
+func getInstanceTypeZoneInfo(ctx context.Context, session *session.Session, region string, types []string, zones []string) (map[string][]string, error) {
```

Comment:
```suggestion
func getInstanceTypeZoneInfo(ctx context.Context, session *session.Session, region string, types []string, zones []string) (map[string]sets.String, error) {
```

Reply:
i had to revert this back to `map[string][]string` because a list of strings is needed to create a new `sets.String`

Reply:
I don't understand your comment.

Reply:
i had committed this suggestion and then realized that `getInstanceTypeZoneInfo` needs to return a a list of strings because `FilterZonesBasedOnInstanceType` and `PreferredInstanceType` use the list to create a new `sets.String` type to act on it. i could make it return `sets.String` so we don't have to call `sets.NewString()`.

Reply:
Yes, that is my suggestion. Return a map of `sets.String` here so that callers don't have to convert the slice into a `sets.String`.

Reply:
done.